### ISSUE TITLE
Signed comparison causing invalid rulenum reference to trigger SEGV

### ISSUE THREAD

**Issue by mcarey42** [opnsense/ports#177]
Hello,
  I'm running into an issue where invalid traffic is being parsed from pflog into the filterlog daemon and causing it to segfault the process.  The triggering event is a mandatory security scan.  The upshot of what is happening is that on line 146 in the file ports/opnsense/filterlog/files/filterlog.c filterlog.c, line 146 (version tag 23.7.1) the comparison should be unsigned.  If not, a rulenr of -1 (0xffffffff) will trigger an invalid array reference and crash the process.  The code is copy/pasted below:

`
			if (rulelabels_max >= (int)rulenr) {
				if (rulelabels[rulenr]) {
					label = rulelabels[rulenr];
				}
			}
			break;
`

When decoding the packet header, I get:
![image](https://github.com/opnsense/ports/assets/2164258/f65dadd2-3755-48af-b51c-061228bda787)

When attached to the core file (and recompiled with symbols, I see:
![image](https://github.com/opnsense/ports/assets/2164258/48c83b21-dfc5-4ab7-913a-fca6df54de7b)

Changing the comparison on line 146 should fix it.  It may, however, introduce other problems.  I only get this error occasionally, and have had no luck writing something to reproduce it.  Some quality time with gdb let me to this conclusion:

`
			if ((u_int32_t)rulelabels_max >= (u_int32_t)rulenr) {
`

Thanks, 
-Mark.

**Comment by fichtner** [comment]
Hi Mark,

Thanks for the report. I'd like to dig a little deeper as `rulenr` comes from the pflog header `hdr` -- I don't think this is related to packet parsing.

Do you still have the core dump or can reproduce? `p *hdr` would be interesting...


Thanks,
Franco

**Comment by fichtner** [comment]
Also this is on 23.7? Not before?

**Comment by Steve4524** [comment]
I work with Mark on the operations side.  We started our migration from pf to OPNsense about a month ago.  Looking back at the logs, we were getting filterlog coredumps last month before the release of 23.7, so this was in 23.1.11.

I have a screenshot from Mark that shows p*hdr:
![image](https://github.com/opnsense/ports/assets/60761276/1e1a36e5-edf8-4298-8d7a-0b83a517e90c)

Thanks.   -Steve

**Comment by fichtner** [comment]
Hi Steve,

Ok, I see a couple of issues, but despite this packet being marked as `action = 0` (PF_PASS) with `reason = 3` (PFRES_SHORT) can you confirm that this packet is not passed? I have a kernel patch coming up that would fix this issue if you could test it.

Also preparing a patch to handle the default `rulenr = -1` which is more or less as OpenBSD implemented it but not even tcpdump where the filterlog code originates from handles this correctly. The crash will go away then and the firewall live log will have a better indication of which rule this caused (although it might need a small patch in core as well).


Cheers,
Franco

**Comment by fichtner** [comment]
Proposed filterlog fix: https://github.com/opnsense/ports/commit/8ca0b1f5726d16468aebfc8053691371bee0f522
Proposed kernel fix: https://github.com/opnsense/src/commit/722bb3dd021332697ee34f271e58a656036e778e

Test kernel:

    # opnsense-update -zkr 23.7.2-pflog

**Comment by Steve4524** [comment]
Franco - First off, super impressed with your responsiveness on this issue, much appreciated.

I've been working on reproducing the crash in our test environment.  Upgraded to 23.7.2, rebooted, ran scan to an IP block across the firewall, filterlog crashed.  Installed test kernel with command above, rebooted, started scan, filterlog crashed.

I'll have to get with Mark to see what's in the latest core file.

Thanks.    -Steve

**Comment by fichtner** [comment]
@Steve4524 thanks a ton for testing this! The `*p hdr` should ideally show `action = 1` in the new core dump. I'll publish a test package for fixed filterlog tomorrow. 


Cheers,
Franco

**Comment by mcarey42** [comment]
Here's the most recent coredump we've manager to induce.  Looks like it's working as intended:  action = 1.

![image](https://github.com/opnsense/ports/assets/2164258/7da5f3b8-231a-4591-a4ee-b983d8430257)

**Comment by fichtner** [comment]
@marjohn56 @Steve4524 thanks guys. here is the filterlog rebuild:

    # opnsense-revert -z filterlog
    # /usr/local/etc/rc.filter_configure


Cheers,
Franco

**Comment by Steve4524** [comment]
@fichtner updated the filterlog package and been running scans this morning.  So far no core dumps, and I'm seeing new types of traffic in the logs.  
```
,,,0,igb0,short,block,in,4,0x0,,64,48239,0,none,6,tcp,20,10.10.10.50,10.254.254.40,0,0,-40,FRPAE,errormsg='[bad hdr length 40 - too long, > 0]',
,,,0,igb0,short,block,in,4,0x0,,64,8823,0,none,1,icmp,20,10.10.10.50,10.254.254.40,datalength=0
,,,0,igb0,short,block,in,4,0x0,,64,603,0,none,17,udp,20,10.10.10.50,10.254.254.40,0,0,31051
```
Only seeing these logs on interface igb0, the segment where the scanner is, which makes me believe these packets are in fact being dropped.

Thanks.    -Steve

**Comment by fichtner** [comment]
@Steve4524 yeah pretty funky bits at the end due to failed parsing. might have to clean this up. I'll feed this to the GUI on my end and see what happens


Cheers,
Franco

**Comment by fichtner** [comment]
Moving this over to core repo to work on the last bits (if required).
